lib/Basic/Utility.cpp

```diff
@@ -0,0 +1,40 @@
+//===-- Utility.cpp -------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2015 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "llbuild/Basic/Utility.h"
+#include <set>
+#include <unordered_map>
+
+namespace llbuild {
+namespace basic {
+
+std::string shellEscaped(StringRef argument) {
```

Comment:
I recommend a couple changes here... they don't have to be done in one PR but it might make the most sense:

1. I would recommend using whitelists not blacklists for characters. So optimize for the common case where the entire string consists only of characters which we know don't need to be escaped, basically `[-a-zA-Z0-9_/:]`(maybe a couple other common ones I forgot). This is the "fast path".

2. If that test fails, I would recommend by probably starting out returning a single-character quoted string. The reason why is that single-character quoted strings have fairly strict shell flexing rules, and so within such strings you only need to quote a very small number of characters (e.g., `'`) -- you can determine if you need any escapes at all during the first fast-path scan.

3. If you only need single quotes, but no escapes, you can just return `' + string + '`, which is the secondary-fast path.

4. If that fails, you do the slow thing of going over the string again character by character and quoting if necessary.